pkg/monitortests/clusterversionoperator/legacycvomonitortests/operators.go

```diff
@@ -35,14 +45,125 @@ func testStableSystemOperatorStateTransitions(events monitorapi.Intervals) []*ju
 			}
 		}
 
-		return "We are not worried about Available=False or Degraded=True blips for stable-system tests yet.", nil
+		isSingleNode, err := isSingleNodeCheck(clientConfig)
+		if err != nil {
+			logrus.Warnf("Error checking for Single Node configuration on stable system (unable to make exception): %v", err)
+			isSingleNode = false
+		}
+
+		// For the non-upgrade case, if any operator has Available=False, fail the test.
+		if condition.Type == configv1.OperatorAvailable && condition.Status == configv1.ConditionFalse {
+			if isSingleNode {
+				return "Operators are allowed to go degraded on single-node for now", nil
+			}
+			if operator == "authentication" {
+				if checkAuthenticationExceptions(condition) {
+					return "https://issues.redhat.com/browse/OCPBUGS-20056", nil
+				}
+			}
+			if operator == "image-registry" {
+				return "Image-registry operator is allowed to have Available=False on a non-upgrade scenario for now", nil
+			}
+			return "", nil
+		}
+		return "We are not worried about Degraded=True blips for stable-system tests yet.", nil
+	}
+
+	return testOperatorStateTransitions(events, []configv1.ClusterStatusConditionType{configv1.OperatorAvailable, configv1.OperatorDegraded}, except, clientConfig)
+}
+
+func isSingleNodeCheck(clientConfig *rest.Config) (bool, error) {
+	configClient, err := clientconfigv1.NewForConfig(clientConfig)
+	if err != nil {
+		logrus.WithError(err).Error("Error creating config client to check for Single Node configuration")
+		return false, err
+	}
+	return exutil.IsSingleNode(context.Background(), configClient)
+}
+
+// isInUpgradeWindow determines if the given eventInterval falls within an upgrade window.
+// UpgradeStart and UpgradeRollback events start upgrade windows and can end and already started upgrade window.
+// UpgradeComplete and UpgradeFailed events end upgrade windows; if there was not an already started upgrade window,
+// we ignore the event.
+// If we don't find any upgrade ending point, we assume the ending point is at the end of the test.
+func isInUpgradeWindow(eventList monitorapi.Intervals, eventInterval monitorapi.Interval) bool {
+	type upgradeWindowHolder struct {
+		startInterval *monitorapi.Interval
+		endInterval   *monitorapi.Interval
+	}
+
+	var upgradeWindows []*upgradeWindowHolder
+	var currentWindow *upgradeWindowHolder
+
+	for _, event := range eventList {
```

Comment:
This looks like we recalculate the upgradeWindows each time and eventInterval is checked.  We could calculate the upgradeWindows one time and then pass them in to compare the eventInterval to cut down on recalculating each time.  That said I thin we only do this when an operator is flagged as not available, question would be how often we see that during upgrades and do we have performance concerns due to that frequency.

Reply:
This is a legitimate call out.  However, given we test probably 35 or so operators and doing that calculation is fast relative to how long these prow jobs run, I don't think performance is a concern.  If someone really wanted to improve the performance, that's fine but I would request we do that in separte PR.